cardano-testnet/src/Testnet/Start/Types.hs

```diff
@@ -99,16 +104,50 @@ instance Default CreateEnvUpdateTime where
   def = CreateEnv
 
 data CreateEnvOptions = CreateEnvOptions
-  { ceoTopologyType :: TopologyType
+  { ceoOnChainParams :: CustomOnChainParams
+  , ceoTopologyType :: TopologyType
   , ceoUpdateTime :: CreateEnvUpdateTime
   } deriving (Eq, Show)
 
 instance Default CreateEnvOptions where
   def = CreateEnvOptions
-    { ceoTopologyType = def
+    { ceoOnChainParams = def
+    , ceoTopologyType = def
     , ceoUpdateTime = def
     }
 
+data CustomOnChainParams
+  = NoCustomParams
+  -- | A file path to a JSON file containing on-chain params, formatted as:
+  -- https://docs.blockfrost.io/#tag/cardano--epochs/GET/epochs/latest
+  | CustomParamsFile FilePath
+  | CustomParamsFlavour OnChainParamsFlavour
+  | CustomParamsGeneses UserProvidedGeneses
+  deriving (Eq, Show)
+
+instance Default CustomOnChainParams where
+  def = NoCustomParams
+
+data UserProvidedGeneses = UserProvidedGeneses
+  { upgShelleyGenesis :: UserProvidedData ShelleyGenesis
+  , upgAlonzoGenesis :: UserProvidedData AlonzoGenesis
+  , upgConwayGenesis :: UserProvidedData ConwayGenesis
+  } deriving (Eq, Show)
+
+instance Default UserProvidedGeneses where
+  def = UserProvidedGeneses
+    def
+    def
+    def
+
+-- | Hard-coded "flavours" for on-chain parameters
+data OnChainParamsFlavour
```

Comment:
I wouldn't bother making this a sum type. Add a constructor called (for example) `MainnetParameters` to `CustomOnChainParams` .